File: backend/src/converters/maa_export.py
"""Export AntennaSim antenna data to MMANA-GAL .maa format.

Generates a .maa file that can be opened in MMANA-GAL, MMANA-GAL basic, etc.
"""

import math
from src.models.antenna import Wire, Excitation, LumpedLoad, LoadType


def export_maa(
    title: str,
    wires: list[Wire],
    excitations: list[Excitation],
    loads: list[LumpedLoad] | None = None,
    frequency_mhz: float = 14.0,
) -> str:
    """Export antenna geometry to MMANA-GAL .maa format.

    Args:
        title: Antenna description/title.
        wires: List of Wire models.
        excitations: List of Excitation models.
        loads: Optional list of LumpedLoad models.
        frequency_mhz: Design frequency in MHz.

    Returns:
        String content of the .maa file.
    """
    lines: list[str] = []
    loads = loads or []

    # Line 0: Title
    lines.append(title or "AntennaSim export")

    # Line 1: Frequency info
    lines.append(f"{frequency_mhz:.6f}")

    # Line 2: Counts: N_wires N_loads N_sources
    lines.append(f"{len(wires)} {len(loads)} {len(excitations)}")

    # Wire geometry lines
    # Format: X1 Y1 Z1 X2 Y2 Z2 Radius N_segments
    for wire in wires:
        lines.append(
            f"{wire.x1:.6f}, {wire.y1:.6f}, {wire.z1:.6f}, "
            f"{wire.x2:.6f}, {wire.y2:.6f}, {wire.z2:.6f}, "
            f"{wire.radius:.6f}, {wire.segments}"
        )

    # Load lines
    # Format: Wire_num Seg_num R X L C
    for load in loads:
        if load.load_type == LoadType.SERIES_RLC:
            lines.append(
                f"{load.wire_tag}, {load.segment_start}, "
                f"{load.param1:.6g}, 0, {load.param2:.6g}, {load.param3:.6g}"
            )
        elif load.load_type == LoadType.FIXED_IMPEDANCE:
            lines.append(
                f"{load.wire_tag}, {load.segment_start}, "
                f"{load.param1:.6g}, {load.param2:.6g}, 0, 0"
            )
        else:
            # Wire conductivity and parallel RLC don't map cleanly to .maa
            lines.append(
                f"{load.wire_tag}, {load.segment_start}, "
                f"{load.param1:.6g}, {load.param2:.6g}, {load.param3:.6g}, 0"
            )

    # Source lines
    # Format: Wire_num Seg_num Voltage_mag Voltage_phase
    for ex in excitations:
        v_mag = math.sqrt(ex.voltage_real ** 2 + ex.voltage_imag ** 2)
        v_phase = math.degrees(math.atan2(ex.voltage_imag, ex.voltage_real))
        lines.append(
            f"{ex.wire_tag}, {ex.segment}, {v_mag:.6f}, {v_phase:.2f}"
        )

    # Ground section (simplified — MMANA format varies)
    lines.append("1")  # Ground type: 1 = real ground
    lines.append("13.0, 0.005")  # Average ground parameters

    # End marker
    lines.append("")

    return "\n".join(lines) + "\n"

File: frontend/src/engine/parsers/maa-export.ts
/**
 * Export antenna geometry to MMANA-GAL .maa format.
 *
 * Generates a .maa file that can be opened in MMANA-GAL, MMANA-GAL basic, etc.
 */

import type { WireGeometry, Excitation } from "../../templates/types";
import type { LumpedLoad } from "../../api/nec";

export function exportMaa(
  title: string,
  wires: WireGeometry[],
  excitations: Excitation[],
  loads?: LumpedLoad[],
  frequencyMhz?: number,
): string {
  const freq = frequencyMhz ?? 14.0;
  const loadList = loads ?? [];
  const lines: string[] = [];

  // Line 0: Title
  lines.push(title || "AntennaSim export");

  // Line 1: Frequency info
  lines.push(freq.toFixed(6));

  // Line 2: Counts: N_wires N_loads N_sources
  lines.push(`${wires.length} ${loadList.length} ${excitations.length}`);

  // Wire geometry lines
  // Format: X1 Y1 Z1 X2 Y2 Z2 Radius N_segments
  for (const wire of wires) {
    lines.push(
      `${wire.x1.toFixed(6)}, ${wire.y1.toFixed(6)}, ${wire.z1.toFixed(6)}, ` +
      `${wire.x2.toFixed(6)}, ${wire.y2.toFixed(6)}, ${wire.z2.toFixed(6)}, ` +
      `${wire.radius.toFixed(6)}, ${wire.segments}`,
    );
  }

  // Load lines
  // Format: Wire_num Seg_num R X L C
  for (const load of loadList) {
    if (load.load_type === 0) {
      // Series RLC
      lines.push(
        `${load.wire_tag}, ${load.segment_start}, ` +
        `${toPrecision(load.param1)}, 0, ${toPrecision(load.param2)}, ${toPrecision(load.param3)}`,
      );
    } else if (load.load_type === 4) {
      // Fixed impedance
      lines.push(
        `${load.wire_tag}, ${load.segment_start}, ` +
        `${toPrecision(load.param1)}, ${toPrecision(load.param2)}, 0, 0`,
      );
    } else {
      // Wire conductivity / parallel RLC — don't map cleanly to .maa
      lines.push(
        `${load.wire_tag}, ${load.segment_start}, ` +
        `${toPrecision(load.param1)}, ${toPrecision(load.param2)}, ${toPrecision(load.param3)}, 0`,
      );
    }
  }

  // Source lines
  // Format: Wire_num Seg_num Voltage_mag Voltage_phase
  for (const ex of excitations) {
    const vMag = Math.sqrt(ex.voltage_real ** 2 + ex.voltage_imag ** 2);
    const vPhaseDeg = (Math.atan2(ex.voltage_imag, ex.voltage_real) * 180) / Math.PI;
    lines.push(
      `${ex.wire_tag}, ${ex.segment}, ${vMag.toFixed(6)}, ${vPhaseDeg.toFixed(2)}`,
    );
  }

  // Ground section (hardcoded: real ground with average parameters)
  lines.push("1");
  lines.push("13.0, 0.005");

  // End marker
  lines.push("");

  return lines.join("\n") + "\n";
}

/** Format a number with up to 6 significant digits, like Python's :.6g */
function toPrecision(n: number): string {
  return Number(n.toPrecision(6)).toString();
}
